src/coprocessor/dag/executor/scanner.rs

```diff
@@ -13,107 +13,105 @@
 
 use kvproto::coprocessor::KeyRange;
 
+use coprocessor::endpoint::prefix_next;
+use coprocessor::codec::table::truncate_as_row_key;
 use storage::{Key, ScanMode, SnapshotStore, Statistics, StoreScanner, Value};
 use storage::txn::Result;
 use util::escape;
+
+#[derive(Copy, Clone)]
+pub enum ScanOn {
+    Table,
+    Index,
+}
+
 // `Scanner` is a helper struct to wrap all common scan operations
 // for `TableScanExecutor` and `IndexScanExecutor`
-pub struct Scanner<'a> {
-    store: SnapshotStore<'a>,
+pub struct Scanner {
     scan_mode: ScanMode,
-    key_only: bool,
-    statistics: Option<&'a mut Statistics>,
-    seek_key: Option<Vec<u8>>,
-    scanner: Option<StoreScanner<'a>>,
+    scan_on: ScanOn,
+    seek_key: Vec<u8>,
+    scanner: StoreScanner,
+    range: KeyRange,
+    no_more: bool,
 }
 
-impl<'a> Scanner<'a> {
+impl Scanner {
     pub fn new(
-        store: SnapshotStore<'a>,
+        store: &SnapshotStore,
+        scan_on: ScanOn,
         desc: bool,
         key_only: bool,
-        statistics: &'a mut Statistics,
-    ) -> Scanner<'a> {
-        let scan_mode = if desc {
-            ScanMode::Backward
+        range: KeyRange,
+    ) -> Result<Scanner> {
+        let (scan_mode, seek_key) = if desc {
+            (ScanMode::Backward, range.get_end().to_vec())
         } else {
-            ScanMode::Forward
+            (ScanMode::Forward, range.get_start().to_vec())
         };
-        Scanner {
-            store: store,
+
+        Ok(Scanner {
             scan_mode: scan_mode,
-            key_only: key_only,
-            statistics: Some(statistics),
-            seek_key: None,
-            scanner: None,
-        }
+            scan_on: scan_on,
+            seek_key: seek_key,
+            scanner: store.scanner(scan_mode, key_only, None)?,
```

Comment:
You need to set `upper_bound` here, otherwise it will take much time to finish if there are too many tombstone keys.

Reply:
Turns out that it's conflict with reusing iterator. Since it's not reusing in the past, we can follow the same way for now.